lib/galaxy/model/security.py

```diff
@@ -1447,62 +1439,171 @@ def get_showable_folders(
+        for role_id in role_ids:
+            stmt = select(Role.id).where(Role.id == role_id).where(Role.type == Role.types.PRIVATE)
+            is_private = bool(self.sa_session.scalars(stmt).all())
+            if not is_private:
+                filtered.append(role_id)
+        return filtered
+
+    def _set_group_users(self, group, user_ids):
+        delete_stmt = delete(UserGroupAssociation).where(UserGroupAssociation.group_id == group.id)
+        insert_values = [{"group_id": group.id, "user_id": user_id} for user_id in user_ids]
+        self._set_associations(group, UserGroupAssociation, delete_stmt, insert_values)
+
+    def _set_group_roles(self, group, role_ids):
+        delete_stmt = delete(GroupRoleAssociation).where(GroupRoleAssociation.group_id == group.id)
+        insert_values = [{"group_id": group.id, "role_id": role_id} for role_id in role_ids]
+        self._set_associations(group, GroupRoleAssociation, delete_stmt, insert_values)
+
+    def _set_role_users(self, role, user_ids):
+        # Do not set users if the role is private
+        # Even though we do not expect to be handling a private role here, the following code is
+        # a safeguard against deleting a user-role-association record for a private role.
+        if role.type == Role.types.PRIVATE:
+            return
+
+        # First, check previously associated users to:
+        # - delete DefaultUserPermissions for users that are being removed from this role;
+        # - delete DefaultHistoryPermissions for histories associated with users that are being removed from this role.
+        for ura in role.users:
+            if ura.user_id not in user_ids:  # If a user will be removed from this role, then:
+                user = self.sa_session.get(User, ura.user_id)
+                # Delete DefaultUserPermissions for this user
+                for dup in user.default_permissions:
+                    if role == dup.role:
+                        self.sa_session.delete(dup)
+                # Delete DefaultHistoryPermissions for histories associated with this user
+                for history in user.histories:
+                    for dhp in history.default_permissions:
+                        if role == dhp.role:
+                            self.sa_session.delete(dhp)
+
+        delete_stmt = delete(UserRoleAssociation).where(UserRoleAssociation.role_id == role.id)
+        insert_values = [{"role_id": role.id, "user_id": user_id} for user_id in user_ids]
+        self._set_associations(role, UserRoleAssociation, delete_stmt, insert_values)
+
+    def _set_role_groups(self, role, group_ids):
+        delete_stmt = delete(GroupRoleAssociation).where(GroupRoleAssociation.role_id == role.id)
+        insert_values = [{"role_id": role.id, "group_id": group_id} for group_id in group_ids]
+        self._set_associations(role, GroupRoleAssociation, delete_stmt, insert_values)
+
+    def _ensure_model_instance_has_id(self, model_instance):
```

Comment:
I would have expected an assertion error or something like this here, not a flush. Maybe rename this to `_flush_if_necessary` or `_persist_new` or something like that.